x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/FallbackableClaim.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc.jwt;
+
+import com.nimbusds.jwt.JWTClaimsSet;
+
+import org.elasticsearch.core.Nullable;
+
+import java.text.ParseException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * A JWT claim that can optionally fallback to another claim (if configured) for retrieving the associated value
+ * from a {@link JWTClaimsSet}. The fallback behaviour happens only when:
+ * 1. The fallback is configured (it can be null)
+ * 2. The original claim does not exist in the {@link JWTClaimsSet}
+ * In any other cases, the original claim will be used for retrieving the value.
+ */
+public class FallbackableClaim {
+    private final String name;
+    private final JWTClaimsSet claimsSet;
+    private final String actualName;
+
+    public FallbackableClaim(String name, @Nullable Map<String, String> fallbackClaimNames, JWTClaimsSet claimsSet) {
+        this.name = Objects.requireNonNull(name);
+        this.claimsSet = Objects.requireNonNull(claimsSet);
+        final String fallbackName;
+        if (fallbackClaimNames != null) {
+            fallbackName = fallbackClaimNames.getOrDefault(name, name);
+        } else {
+            fallbackName = null;
+        }
+        if (fallbackName == null) {
+            this.actualName = name;
+        } else {
+            this.actualName = claimsSet.getClaim(name) != null ? name : fallbackName;
+        }
+    }
+
+    public String getActualName() {
+        return actualName;
+    }
+
+    public String getStringClaimValue() {
+        try {
+            return claimsSet.getStringClaim(actualName);
+        } catch (ParseException e) {
+            throw new IllegalArgumentException(format("cannot parse string claim [%s] as string", this), e);
+        }
+    }
+
+    public List<String> getStringListClaimValue() {
+        final Object claimValue = claimsSet.getClaim(actualName);
+        if (claimValue instanceof String) {
+            return List.of((String) claimValue);
+        } else {
+            try {
+                return claimsSet.getStringListClaim(actualName);
+            } catch (ParseException e) {
+                throw new IllegalArgumentException(format("cannot parse string claim [%s] as string array", this), e);
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        if (name.equals(actualName)) {
+            return name;
+        } else {
+            return format("%s (fallback of %s)", actualName, name);
```

Comment:
```suggestion
            return format("%s (with fallback to %s)", actualName, name);
```

"fallback of" feels ambiguous as to which claim-name is which.

Reply:
As commented above, the `toString` method produces different strings depending on whether the fallback claim is actually used. This string is for when the fallback claim is effective, i.e. the main claim does not exist. 

IIUC, your suggestion is more of an uniform string representation of a `FallbackableClaim`, i.e. it always shows `main_claim (with fallback to another_claim)`. This is probably better consistency but lacks the ability to tell which claim is actually used and would require yet another string representation for that. I prefer to have an "adaptable" `toString` method so the object can be used by itself in logging and exception messages and be able to give sufficient info about its internal state. Therefore I'll keep it as is for now. Please let me know if you still have a different view and we can iterate on it in a future PR.